azure/services/subnets/spec.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnets
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+)
+
+// SubnetSpec defines the specification for a Subnet.
+type SubnetSpec struct {
+	Name              string
+	ResourceGroup     string
+	SubscriptionID    string
+	CIDRs             []string
+	VNetName          string
+	IsVNetManaged     bool
+	RouteTableName    string
+	SecurityGroupName string
+	Role              infrav1.SubnetRole
+	NatGatewayName    string
+}
+
+// ResourceName returns the name of the subnet.
+func (s *SubnetSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *SubnetSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the VNet that owns this subnet.
+func (s *SubnetSpec) OwnerResourceName() string {
+	return s.VNetName
+}
+
+// Parameters returns the parameters for the subnet.
+func (s *SubnetSpec) Parameters(existing interface{}) (parameters interface{}, err error) {
+	if existing != nil {
+		_, ok := existing.(network.Subnet)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.Subnet", existing)
+		}
+
+		return nil, nil
+	}
+
+	if !s.IsVNetManaged {
+		// TODO: change this to terminal error once we add support for handling them
+		return nil, errors.Errorf("custom vnet was provided but subnet %s is missing", s.Name)
+	}
+	subnetProperties := network.SubnetPropertiesFormat{
+		AddressPrefixes: &s.CIDRs,
+	}
+
+	// workaround needed to avoid SubscriptionNotRegisteredForFeature for feature Microsoft.Network/AllowMultipleAddressPrefixesOnSubnet.
+	if len(s.CIDRs) == 1 {
+		subnetProperties = network.SubnetPropertiesFormat{
+			AddressPrefix: &s.CIDRs[0],
+		}
+	}
+
+	if s.RouteTableName != "" {
+		subnetProperties.RouteTable = &network.RouteTable{
+			ID: to.StringPtr(azure.RouteTableID(s.SubscriptionID, s.ResourceGroup, s.RouteTableName)),
```

Comment:
This will point to vnet's resource group, but we need cluster's resource group here. We'll need to also set cluster's resource group in SubnetSpec and use it construct sub resource ids. Perhaps something like https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/dc0c3fad698c7c788ae5a14cfbb8c3c94180f97b/azure/services/loadbalancers/spec.go#L39

Reply:
Do we really need the cluster RG here? There are two cases:
1) the vnet is created by CAPZ: then the vnet RG _is_ the cluster RG 
2) the vnet is custom: then we don't create the subnet, so this doesn't matter and the subnet should already have the right RT, NAT GW and NSG attached to it)

Reply:
Yeah that's what I was thinking. I'm not really sure why we'd expect the route tables, nat gateways, and security groups to be in a separate RG if the vnet is custom.

Reply:
can we not customize vnet rg even when it is managed by capz? iiuc, it's only set to cluster rg if it's empty https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5c96e824f739bc8627cab4017312e2054d329d63/api/v1beta1/azurecluster_default.go#L77

Reply:
I'm not sure what you mean, if we specify the VNet RG here then the value should be set in `s.Vnet().ResourceGroup` right?

Reply:
Yes, but it wont be the same as cluster rg, right? The current implementation also suggest this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5c96e824f739bc8627cab4017312e2054d329d63/azure/services/subnets/subnets.go#L84

Reply:
Perhaps it's a bug in the current implementation. But, right now, the way it works is:
1. Vnet and Subnets are created in the virtual network's resource group - `NetworkSpec.Vnet.ResourceGroup`
2. All other resources (route tables, nat gateways, security groups) are created in the cluster resource group - `AzureCluster.ResourceGroup`

Reply:
@shysank is correct, I did not think of the case where the user doesn't create the vnet but sets the RG in the spec to be different than the cluster RG. In that case we need to use cluster RG for these since it's not always the same as vnet RG.

Reply:
@shysank i believe the reasoning is that you can share a vnet between multiple clusters (and so it makes sense to put it in a common place like a networking RG). Subnets have to be in the same RG as the vnet since they are a nested resource. As for everything else, it’s a 1:1 relationship with the cluster hence putting it all in the cluster RG. If this is unexpected we could rework that UX. Let’s tackle that in a separate issue though and keep the behavior the same for this PR though.